### PR TITLE
Channel message dispatch only when it's attached

### LINKED ISSUE #240: RTL17: Only dispatch messages when ATTACHED
Write spec tests for [RTL17](https://docs.ably.io/client-lib-development-guide/features/#RTL17), ensure they pass and that the affected API conforms to [the 1.2 API design](https://github.com/ably/ably-go/issues/197).

### PR DESCRIPTION
**Pull request by sacOO7**
- Added missing channel attached check
- Updated tests for the same
- Fixes #240

### CODE REVIEW

**Review by tcard — APPROVED** [review]

File: ably/realtime_channel.go [hunk 1]
```diff
@@ -563,8 +563,10 @@ func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 		c.setState(ChannelStateFailed, newErrorFromProto(msg.Error))
 		c.queue.Fail(newErrorFromProto(msg.Error))
 	case proto.ActionMessage:
-		for _, msg := range msg.Messages {
-			c.messageEmitter.Emit(subscriptionName(msg.Name), (*subscriptionMessage)(msg))
+		if c.state == ChannelStateAttached {
```
Comment (tcard) [inline ably/realtime_channel.go:566]: This access to `c.state` should happen while holding a lock on `c.mtx`. (This really should be easier to deal with; see #184.)

```suggestion
		c.mtx.Lock()
		state := c.state
		c.mtx.Unlock()
		if state == ChannelStateAttached {
```

**Review by tcard — CHANGES_REQUESTED** [review]

**Comment by tcard** [comment]
It looks like there's a test that's consistently failing:

```
=== CONT  TestRealtimeChannel_Detach
    realtime_channel_test.go:151: waiting on subscribed channel close timed out after 30s
```

**Comment by sacOO7** [comment]
> It looks like there's a test that's consistently failing:
> 
> ```
> === CONT  TestRealtimeChannel_Detach
>     realtime_channel_test.go:151: waiting on subscribed channel close timed out after 30s
> ```

Yeah, seems it's flaky, I am not sure if there is a way to fix it.

**Review by lmars — APPROVED** [review]

**Comment by lmars** [comment]
@sacOO7 is there anything remaining to be done on this PR?

**Pull request merged**